### PR TITLE
[#11219] Fix autoscrolling when viewing session result

### LINKED ISSUE #11219: Autoscrolling doesn't reach the correct question when moderating response
- **Environment**: `master` branch at commit https://github.com/TEAMMATES/teammates/commit/eb6f5dba9927a7149cf2a52f735199659374781a

**Steps to reproduce**
1. Instructor view a session result.
2. Instructor wants to moderate a response.

**Expected behaviour**

https://user-images.githubusercontent.com/58477697/122939239-0c571e80-d3a6-11eb-90f4-9f6a97b1030c.mp4

**Actual behaviour**
It started to have problem from question 3 onward

https://user-images.githubusercontent.com/58477697/122939609-5b04b880-d3a6-11eb-92b1-396647de6199.mp4

### PR DESCRIPTION
**Pull request by tjtanjin**
Fixes #11219

**Outline of Solution**

Add checks to ensure that scrolling continues until the below conditions are satisfied:
- Target to scroll to is not null
- Target to scroll to is visible on screen

An example of the autoscroll for question 11 is shown below:
https://user-images.githubusercontent.com/43908963/123521950-88c26800-d6ec-11eb-9e3a-cddaab7a4973.mov

Note that although the scrolling in the video stops at question 9, the question itself still matches with the original question 11 that was clicked. The reason for this discrepancy is because 2 of the earlier questions have no responses and were not displayed (not sure if this might actually be confusing/misleading).

### CODE REVIEW

**Comment by jianhandev** [comment]
Hi @tjtanjin it looks like your PR is failing some checks, you can refer to [this document](https://github.com/TEAMMATES/teammates/blob/master/docs/static-analysis.md) on how to run static analysis. Make sure to pass the [component tests](https://github.com/TEAMMATES/teammates/blob/master/docs/development.md#testing) and [e2e tests](https://github.com/TEAMMATES/teammates/blob/master/docs/e2e-testing.md) as well.

File: src/web/app/pages-session/session-submission-page/session-submission-page.component.ts [hunk 1]
```diff
@@ -203,17 +203,47 @@ export class SessionSubmissionPageComponent implements OnInit, AfterViewInit {
     });
   }
 
+  // Solution for checking partial element visibility adapted from
+  // https://www.codegrepper.com/code-examples/javascript/read+xlsx+and+xls+file+in+typescript
```
Comment (t-cheepeng) [inline src/web/app/pages-session/session-submission-page/session-submission-page.component.ts:207]: The link is wrong? It brings me to a code snippet on opening excel files

Comment (tjtanjin) [reply src/web/app/pages-session/session-submission-page/session-submission-page.component.ts]: Thanks for pointing that out! I've fixed the link along with the errors for passing checks! :)

**Review by t-cheepeng — APPROVED** [review]
LGTM

File: src/web/app/pages-session/session-submission-page/session-submission-page.component.ts [hunk 2]
```diff
@@ -203,17 +203,47 @@ export class SessionSubmissionPageComponent implements OnInit, AfterViewInit {
...
+      Math.floor(100 - (((rect.top >= 0 ? 0 : rect.top) / +-rect.height) * 100)) < 1 ||
+      Math.floor(100 - ((rect.bottom - windowHeight) / rect.height) * 100) < 1
+    );
+  }
+
+  /**
+   * Scrolls to the question based on its given question id.
+   */
+  scrollToQuestion(): void {
+    const div: any = document.getElementById(this.moderatedQuestionId);
```
Comment (ccyccyccy) [inline src/web/app/pages-session/session-submission-page/session-submission-page.component.ts:227]: Don't have to type as `any` here

File: src/web/app/pages-session/session-submission-page/session-submission-page.component.ts [hunk 3]
```diff
@@ -203,17 +203,47 @@ export class SessionSubmissionPageComponent implements OnInit, AfterViewInit {
...
+
+    // continue scrolling as long as the element to scroll to is yet to be found or not in view
+    if (div == null || !(this.isInViewport(div))) {
```
Comment (ccyccyccy) [inline src/web/app/pages-session/session-submission-page/session-submission-page.component.ts:230]: Use `===` instead of `==`

**Review by ccyccyccy — DISMISSED** [review]
Approach looks good to me.
Can post a GIF or video here as verification? Thanks!

File: src/web/app/pages-session/session-submission-page/session-submission-page.component.ts (hunk 2 above)
Comment (tjtanjin) [reply src/web/app/pages-session/session-submission-page/session-submission-page.component.ts:227]: I initially went with type as `HTMLElement` but it resulted in this error message in instances where the element is null: `Type 'null' is not assignable to type 'HTMLElement'` - hence the choice of `any`. Are there alternatives to this?

**Comment by rrtheonlyone** [comment]
@tjtanjin kind reminder on this PR

File: src/web/app/pages-session/session-submission-page/session-submission-page.component.ts (hunk 2 above)
Comment (tjtanjin) [reply src/web/app/pages-session/session-submission-page/session-submission-page.component.ts:227]: Hello @ccyccyccy! With reference to my comment above, would you have any advice or is it alright to keep the type `any`? >.< Thank You!

Comment (rrtheonlyone) [reply src/web/app/pages-session/session-submission-page/session-submission-page.component.ts:227]: Hi @tjtanjin why do you need to type it like that? Typescript should be able to infer the type: https://www.typescriptlang.org/docs/handbook/type-inference.html

**Review by rrtheonlyone — APPROVED** [review]
LGTM 💯 

Good solution here - works

**Pull request merged**